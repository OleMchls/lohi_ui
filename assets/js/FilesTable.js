import React, { Component } from 'react'
import { library } from '@fortawesome/fontawesome-svg-core'
import { fas } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

library.add(fas)

const FilesTable = ({ files, onMove, onRemove }) => {
  if (!files.length) return null

  return (
    <table className="table table-striped">
      <thead>
        <tr>
          <th>#</th>
          <th>File</th>
          <th>Actions</th>
        </tr>
      </thead>
      <tbody>
        {files.map((file, idx) => (
          <tr key={file.uuid}>
            <th scope="row">{idx + 1}</th>
            <td>{file.name}</td>
            <td>
              {idx !== 0 && (
                <button
                  title={`move ${file.name} up`}
                  type="button"
                  className="btn btn-link"
                  onClick={() => onMove(idx, idx - 1)}
                >
                  <FontAwesomeIcon icon="chevron-up" aria-hidden="true" />
                </button>
              )}
              {idx !== files.length - 1 && (
                <button
                  title={`move ${file.name} down`}
                  type="button"
                  className="btn btn-link"
                  onClick={() => onMove(idx, idx + 1)}
                >
                  <FontAwesomeIcon icon="chevron-down" aria-hidden="true" />
                </button>
              )}
              <button
                title={`remove ${file.name}`}
                type="button"
                className="btn btn-link"
                onClick={() => onRemove(file)}
              >
                <FontAwesomeIcon icon="minus" aria-hidden="true" />
              </button>
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}

export default FilesTable
