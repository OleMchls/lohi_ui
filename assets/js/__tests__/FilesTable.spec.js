import React from 'react'
import { render, fireEvent, cleanup } from 'react-testing-library'
import FilesTable from '../FilesTable'

const renderTable = (overrides = {}) => {
  const props = {
    files: [{ uuid: 1, name: 'file_1' }, { uuid: 2, name: 'file_2' }],
    onMove: jest.fn().mockName('onMove'),
    onRemove: jest.fn().mockName('onRemove'),
    ...overrides
  }

  return { ...render(<FilesTable {...props} />), props }
}

afterEach(cleanup)

it('renders null with no files provided', () => {
  const { container } = renderTable({ files: [] })

  expect(container).toBeEmpty()
})

it('renders a table with 2 files', () => {
  const { container } = renderTable()

  expect(container).toMatchSnapshot()
})

it('moves an item up and down', () => {
  const {
    getByTitle,
    props: { onMove }
  } = renderTable()

  fireEvent.click(getByTitle('move file_1 down'))
  expect(onMove).toHaveBeenCalledWith(0, 1)

  fireEvent.click(getByTitle('move file_2 up'))
  expect(onMove).toHaveBeenCalledWith(1, 0)
})

it('deletes an item', () => {
  const {
    getByTitle,
    props: { onRemove }
  } = renderTable()

  fireEvent.click(getByTitle('remove file_1'))

  expect(onRemove).toHaveBeenCalledWith({ uuid: 1, name: 'file_1' })
})
