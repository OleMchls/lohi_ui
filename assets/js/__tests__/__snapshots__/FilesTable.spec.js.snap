// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a table with 2 files 1`] = `
<div>
  <table
    class="table table-striped"
  >
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          File
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          1
        </th>
        <td>
          file_1
        </td>
        <td>
          <button
            class="btn btn-link"
            title="move file_1 down"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down fa-w-14 "
              data-icon="chevron-down"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                fill="currentColor"
              />
            </svg>
          </button>
          <button
            class="btn btn-link"
            title="remove file_1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-minus fa-w-14 "
              data-icon="minus"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          2
        </th>
        <td>
          file_2
        </td>
        <td>
          <button
            class="btn btn-link"
            title="move file_2 up"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-up fa-w-14 "
              data-icon="chevron-up"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
                fill="currentColor"
              />
            </svg>
          </button>
          <button
            class="btn btn-link"
            title="remove file_2"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-minus fa-w-14 "
              data-icon="minus"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
